### PR TITLE
Add export to public API

### PR DESCRIPTION
**Pull request by petr-hanzl**
This PR adds option to export LiveDB genesis for certain block from Archive while LiveDB is syncing.

### CODE REVIEW

File: go/carmen/carmen.go [hunk 1]
```diff
@@ -198,6 +200,11 @@ type HistoricBlockContext interface {
 	// otherwise, the proof is returned.
 	GetProof(address Address, keys ...Key) (WitnessProof, error)
 
+	// CreateLiveDBGenesis writes a LiveDB genesis file for given block into out.
+	// The genesis file is exported from the archive for the
+	// given block and is used to sync LiveDB to that block.
+	CreateLiveDBGenesis(out io.Writer) (Hash, error)
```
Comment (HerbertJordan) [inline go/carmen/carmen.go:206]: I think the term `Genesis` should not be in this function name. While it is true that what gets exported is potentially part of a genesis file, it is only a small part of it, and hardly the defining element.

I would suggest to rename this function to something like`Export()`. Wdyt?

File: go/backend/archive/archive.go [hunk 2]
```diff
@@ -60,6 +65,9 @@ type Archive interface {
 	// CreateWitnessProof creates a witness proof for the given account and keys.
 	CreateWitnessProof(block uint64, address common.Address, keys ...common.Key) (witness.Proof, error)
 
+	// CreateLiveDBGenesis writes a LiveDB genesis for given block into out.
+	CreateLiveDBGenesis(block uint64, out io.Writer) (common.Hash, error)
```
Comment (HerbertJordan) [inline go/backend/archive/archive.go:69]: As in the Carmen interface, I would argue that what is exported is not a `Genesis`.

File: go/carmen/carmen.go (hunk 1 above)
Comment (HerbertJordan) [inline go/carmen/carmen.go:206]: This function will eventually also need a `context.Context` to support canceling the export, right?

File: go/carmen/example_test.go [hunk 3]
```diff
@@ -515,3 +518,82 @@ func ExampleDatabase_GetMemoryFootprint() {
...
+
+	// Export genesis
+	var rootHash carmen.Hash
+	b := bytes.NewBuffer(nil)
+	if err = db.QueryBlock(uint64(1), func(ctxt carmen.HistoricBlockContext) error {
+		rootHash, err = ctxt.CreateLiveDBGenesis(b)
+		if err != nil {
+			log.Fatalf("cannot create genesis: %v", err)
+		}
+		return nil
```
Comment (HerbertJordan) [inline go/carmen/example_test.go:559]:
```suggestion
		return err
```

File: go/carmen/example_test.go [hunk 4]
```diff
@@ -515,3 +518,82 @@ func ExampleDatabase_GetMemoryFootprint() {
...
+	}); err != nil {
+		log.Fatalf("cannot query block: %v", err)
```
Comment (HerbertJordan) [inline go/carmen/example_test.go:561]:
```suggestion
		log.Fatalf("failed to export block: %v", err)
```

File: go/common/genesis/genesis.go [hunk 5]
```diff
@@ -8,7 +8,7 @@
 // On the date above, in accordance with the Business Source License, use of
 // this software will be governed by the GNU Lesser General Public License v3.
 
-package io
+package genesis
```
Comment (HerbertJordan) [inline go/common/genesis/genesis.go:11]: Not a real fan of the `genesis` term due to its much wider (and to some extend different) meaning in the context of a block chain.

If this package is needed (see other comments) we should think about a better name.

File: go/common/genesis/genesis.go
Comment (HerbertJordan) [inline go/common/genesis/genesis.go:1]: Why do those formally private constants, types, and functions have to be made public?

File: go/common/genesis/genesis.go [hunk 6]
```diff
@@ -38,7 +51,7 @@ func writeCodes(codes map[common.Hash][]byte, out io.Writer) error {
 	return nil
 }
 
-func readCode(in io.Reader) ([]byte, error) {
+func ReadCode(in io.Reader) ([]byte, error) {
```
Comment (HerbertJordan) [inline go/common/genesis/genesis.go:54]: This function used to be a utility function for a very specific purpose. By making it public and putting it in a common package, much more documentation would be required to describe what this function is supposed to do.

File: go/database/mpt/export.go
Comment (HerbertJordan) [inline go/database/mpt/export.go:1]: There is no need to move this code from [io/live.go](https://github.com/Fantom-foundation/Carmen/blob/main/go/database/mpt/io/live.go#L114) to this location. Keeping it where it was would eliminate the need for the new `genesis` package, which's purpose is not well defined.

File: go/database/mpt/io/live.go [hunk 7]
```diff
@@ -127,49 +83,8 @@ func ExportBlockFromArchive(ctx context.Context, directory string, out io.Writer
...
-	}
-
-	// Write out all accounts and values.
-	visitor := exportVisitor{out: out, ctx: ctx}
-	if err := db.Visit(&visitor); err != nil || visitor.err != nil {
-		return fmt.Errorf("failed exporting content: %w", errors.Join(err, visitor.err))
-	}
-
-	return nil
+	_, err = mpt.ExportLive(ctx, mpt.NewExportableArchiveTrie(archive, block), out)
```
Comment (HerbertJordan) [inline go/database/mpt/io/live.go:86]: Why is this moved to the `mpt` package?

File: go/state/cppstate/cpp_state.go [hunk 8]
```diff
@@ -285,6 +287,10 @@ func (cs *CppState) CreateWitnessProof(address common.Address, keys ...common.Ke
 	panic("not implemented")
 }
 
+func (cs *CppState) CreateLiveDBGenesis(out io.Writer) (common.Hash, error) {
+	panic("not implemented")
```
Comment (HerbertJordan) [inline go/state/cppstate/cpp_state.go:291]: Should return an error, not panic.

File: go/state/gostate/go_state.go [hunk 9]
```diff
@@ -368,6 +370,10 @@ func (s *GoState) GetProof() (backend.Proof, error) {
 	return backend.GetComposedProof(proofs), nil
 }
 
+func (s *GoState) CreateLiveDBGenesis(io.Writer) (common.Hash, error) {
+	panic("not implemented")
```
Comment (HerbertJordan) [inline go/state/gostate/go_state.go:374]: Prefer errors over `panic`.

**Review by HerbertJordan — CHANGES_REQUESTED** [review]
I think some of the code movements of export functions are adding additional complexity that should be avoided (in particular the need for the `common/genesis` package).

Before this PR, exporting a dump for a single world state from an archive was already introduced by #966, and it is working. It is surprising that the need make this public available through the Carmen facade causes so much refactoring of the already existing export infrastructure.

Furthermore, I think we should avoid the term `Genesis` altogether since what we are exporting is not a `Genesis` dataset, merely a small part of it.

File: go/carmen/carmen.go (hunk 1 above)
Comment (petr-hanzl) [reply go/carmen/carmen.go:206]: You are right, this is only part of the whole `genesis` file. I will remove the term from the whole PR and simply use `Export()`.

File: go/database/mpt/io/live.go [hunk 10]
```diff
@@ -54,9 +54,16 @@ const (
...
+	return exportableArchiveTrie{
+		trie:  trie,
+		block: block,
+	}
+}
+
+// MptStateVisitor is an interface for Tries that allows for visiting the Trie nodes
 // and furthermore getting its properties such as a root hash and contract codes.
-type mptStateVisitor interface {
+type MptStateVisitor interface {
```
Comment (HerbertJordan) [inline go/database/mpt/io/live.go:66]: There is not really a need to make this interface public.

If this interface is going to be public, I have higher quality expectations on its name.
The thing this interfaces defined is not a `Visitor`, so it shout not be named this way.

How about `VisitableState` or `ExportableState`?

**Review by HerbertJordan — CHANGES_REQUESTED** [review]
Some minor requests for improvements; see coments.

File: go/database/mpt/io/live.go [hunk 11]
```diff
@@ -54,9 +54,16 @@ const (
 	EthereumHash = HashType(0)
 )
 
-// mptStateVisitor is an interface for Tries that allows for visiting the Trie nodes
+func NewExportableArchiveTrie(trie *mpt.ArchiveTrie, block uint64) MptStateVisitor {
```
Comment (HerbertJordan) [inline go/database/mpt/io/live.go:57]: Needs documentation.

File: go/database/mpt/io/live.go (hunk 10 above)
Comment (petr-hanzl) [reply go/database/mpt/io/live.go:66]: made private

File: go/carmen/example_test.go [hunk 12]
```diff
@@ -515,3 +519,81 @@ func ExampleDatabase_GetMemoryFootprint() {
...
+	}
+
+	if err := db.Close(); err != nil {
+		log.Fatalf("cannot close db: %v", err)
+	}
+	if err := os.RemoveAll(dir); err != nil {
+		log.Fatalf("cannot remove dir: %v", err)
+	}
+
+	importedDbPath := filepath.Join(os.TempDir(), "imported_carmen")
```
Comment (HerbertJordan) [inline go/carmen/example_test.go:571]: This can lead to collisions with already existing copies of previous runs. Better to use
```
os.MkdirTemp("", "carmen_db_*")
```
like in the first line of this test.

**Review by HerbertJordan — DISMISSED** [review]

**Review by HerbertJordan — APPROVED** [review]
